packages/adapters-library/src/SQLiteMetadataProvider.ts

```diff
@@ -43,109 +45,106 @@ export type PoolRow = {
+    t.token_decimals AS main_token_decimals,
+    ut.token_address AS underlying_token_address,
+    ut_t.token_name AS underlying_token_name,
+    ut_t.token_symbol AS underlying_token_symbol,
+    ut_t.token_decimals AS underlying_token_decimals,
+    ut.additional_data AS underlying_additional_data,
+    rt.token_address AS reward_token_address,
+    rt_t.token_name AS reward_token_name,
+    rt_t.token_symbol AS reward_token_symbol,
+    rt_t.token_decimals AS reward_token_decimals,
+    rt.additional_data AS reward_additional_data,
+    ert.token_address AS extra_reward_token_address,
+    ert_t.token_name AS extra_reward_token_name,
+    ert_t.token_symbol AS extra_reward_token_symbol,
+    ert_t.token_decimals AS extra_reward_token_decimals,
+    ert.additional_data AS extra_reward_additional_data
+  FROM  adapters a 
+        LEFT JOIN pools p ON a.adapter_id = p.adapter_id
+        LEFT JOIN tokens t ON p.pool_address = t.token_address
+        LEFT JOIN underlying_tokens ut ON p.pool_id = ut.pool_id
+        LEFT JOIN tokens ut_t ON ut.token_address = ut_t.token_address
+        LEFT JOIN reward_tokens rt ON p.pool_id = rt.pool_id
+        LEFT JOIN tokens rt_t ON rt.token_address = rt_t.token_address
+        LEFT JOIN extra_reward_tokens ert ON p.pool_id = ert.pool_id
+        LEFT JOIN tokens ert_t ON ert.token_address = ert_t.token_address;
+`
+
 export class SQLiteMetadataProvider implements IMetadataProvider {
-  database: DbType
+  database: Database.Database
+
+  allTokens: Promise<Map<string, ProtocolToken[]>>
 
   constructor(filename: string | Buffer, options: Database.Options) {
     this.database = new Database(filename, options)
 
-    this.database.pragma('journal_mode = DELETE')
-  }
-
-  getMetadata(input: {
-    chainId: Chain
-    protocolId: Protocol
-    productId: string
-  }): Promise<ProtocolToken[]> {
-    return this.getPoolsFromDb(input)
-  }
+    this.allTokens = new Promise<Map<string, ProtocolToken[]>>(
+      (resolve, reject) => {
+        try {
+          const allRows = this.database
+            .prepare(selectAllPoolsQuery)
+            .all() as PoolRow[]
 
-  async getPoolCount(protocolId: string, productId: string): Promise<number> {
-    // Prepare the query to get the adapter ID and count the pools
-    const query = `
-      SELECT COUNT(*) AS pool_count
-      FROM pools
-      WHERE adapter_id = (
-        SELECT adapter_id FROM adapters WHERE protocol_id = ? AND product_id = ?
-      );
-    `
-
-    // Execute the query with the provided protocolId and productId
-    const stmt = this.database.prepare(query)
-    const result = stmt.get(protocolId, productId) as { pool_count?: number }
-
-    // Return the pool count from the result
-    return Number(result.pool_count ?? 0)
+          resolve(this.formatDbResponse(allRows))
+        } catch (error) {
+          logger.error({ filename, options }, 'Error fetching database rows')
+          reject(error)
+        }
+      },
+    )
```

Comment:
We start fetching the data when the provider is instantiated, not after the first query. What we store is a promise, so when adapters try to fetch the data they just need to await it, but it should already be there as it's a very fast query.